<a href="https://colab.research.google.com/github/MarynaZinchenko888/python_for_ds_tasks/blob/main/Zinchenko_M_%22HW_%D0%A2%D0%B5%D1%81%D1%82%D0%B8_%D0%B4%D0%BB%D1%8F_%D0%BC%D0%B0%D0%BB%D0%B8%D1%85_%D0%B2%D0%B8%D0%B1%D1%96%D1%80%D0%BE%D0%BA_%D1%82%D0%B0_%D0%BF%D1%80%D0%BE%D0%BF%D0%BE%D1%80%D1%86%D1%96%D0%B9_ipynb%22.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Тести для малих вибірок та пропорцій



**Завдання 1**. E-commerce компанія після редизайну сайту підозрює, що **середній час до покупки (time-to-purchase)** користувачів **збільшився**.
Історично середній час від першого заходу на сайт до покупки становив $\mu_0 = 123.7$ хвилин.

Команда аналітиків випадково обрала дані **7 покупців після редизайну**:
`128, 135, 121, 142, 126, 151, 123`

З допомогою наявних даних зʼясуйте, чи збільшився середній час покупки після редизайну.

Для цього виконайте наступні 6 кроків. Правильне виконання кожного з кроків оцінюється в 1 бал.

1. Запишіть нульову та альтернативну гіпотези і визначте тип тесту.
2. Обчисліть вибіркові статистики: $\bar x$, $s$, $n$.
3. Оберіть тип тесту та виконайте його (знайдіть критичне значення тесту, статистику тесту та р-значення) будь-яким пасуючим способом, щоб перевірити гіпотезу на рівні значущості $\alpha = 0.10$.
4. Прийміть рішення, чи відхиляєте ви гіпотезу $H_0$ використовуючи p-value.
5. Напишіть висновок: чи справді редизайн сайту подовжив час до покупки?
6. Чи зміниться ваше рішення при зміні рівня значущості на $\alpha = 0.05$.


In [12]:
import numpy as np
from scipy import stats

# Дані
data = np.array([128, 135, 121, 142, 126, 151, 123])
mu0 = 123.7   # історичне середнє
alpha = 0.10

# 1. Вибіркові статистики
n = len(data)
x_bar = np.mean(data)
s = np.std(data, ddof=1)

# 2. t-статистика вручну
t_stat = (x_bar - mu0) / (s / np.sqrt(n))
df = n - 1

# 3. Критичне значення та p-value
t_crit = stats.t.ppf(1 - alpha, df)
p_value = 1 - stats.t.cdf(t_stat, df)  # одностороннє p-value

In [13]:
print("n =", n)
print("x̄ =", round(x_bar, 2))
print("s =", round(s, 2))
print("t-stat =", round(t_stat, 3))
print("df =", df)
print("t-critical (α=0.10) =", round(t_crit, 3))
print("p-value =", round(p_value, 4))

n = 7
x̄ = 132.29
s = 10.98
t-stat = 2.069
df = 6
t-critical (α=0.10) = 1.44
p-value = 0.042


In [14]:
# 4. Висновок
if p_value < alpha:
    print("Відхиляємо H0: час до покупки збільшився (статистично значущо).")
else:
    print("Не відхиляємо H0: немає доказів, що час до покупки збільшився.")

Відхиляємо H0: час до покупки збільшився (статистично значущо).


**Завдання 2.**

До спеціальної рекламної кампанії **23%** дорослих упізнавали логотип компанії. Після завершення кампанії відділ маркетингу провів опитування: з **1200** випадково відібраних дорослих **311** упізнали логотип.

Перевірте на рівні значущості **3%** ($\alpha=0.03$), чи дають ці дані достатні підстави стверджувати, що **тепер більше ніж 23%** дорослих упізнають логотип компанії. Для розвʼязку використовуйте бібліотеку `statsmodels`.

Зробіть висновок, чи зросла впізнаваність логотипу.

Додатково, обчисліть довірчий інтревал на заданому рівні значущості і проінтерпретуйте текстом - як він додатково пояснює прийняте нами рішення?

In [15]:
import statsmodels.api as sm
from statsmodels.stats.proportion import proportions_ztest, proportion_confint

count = 311   # кількість "успіхів" - упізнали логотип
nobs = 1200    # обсяг вибірки
p0 = 0.23      # гіпотетична пропорція (H0)

# z-тест для пропорції
stat, pval = proportions_ztest(count, nobs, p0, alternative='larger')
confint = proportion_confint(count, nobs, alpha=0.03, method='normal')

In [16]:
print("🔹 Тест для однієї пропорції")
print(f"Z-статистика = {stat:.3f}")
print(f"p-value = {pval:.4f}")
print(f"97% довірчий інтервал: {confint}")

🔹 Тест для однієї пропорції
Z-статистика = 2.306
p-value = 0.0106
97% довірчий інтервал: (0.23171700302179205, 0.28661633031154127)


Інтерпретація результатів.


1. Односторонній тест

  * $H_0:p=0.23$
  * $H_1:p>0.23$
  * Оскільки $p\text{-value} = 0.0106<0.03$, відхиляємо  $H_0$.
    👉 Є статистично значущі докази стверджувати, що частка тих, хто упізнає логотип, **більша за 23%**.
2. 97% довірчий інтервал (0.232, 0.287)

  Це означає, що з імовірністю 97% справжня частка лежить у межах від **23,2% до 28,7%**.